src/Lykke.Job.TransactionHandler/Handlers/EthereumCoreCommandHandler.cs

```diff
@@ -0,0 +1,468 @@
+                }
+
+                if (createPendingActions)
+                {
+                    if (asset.IsTrusted)
+                    {
+                        await _ethererumPendingActionsRepository.CreateAsync(clientId, Guid.NewGuid().ToString());
+                    }
+                }
+
+                var result = await _matchingEngineClient.CashInOutAsync(cashinId, clientId, asset.Id, (double)amount);
+
+                if (result == null ||
+                   (result.Status != MeStatusCodes.Ok &&
+                    result.Status != MeStatusCodes.Duplicate))
+                {
+                    _log.WriteWarning(command.TransactionHash ?? "Empty", result, "ME error");
+
+                    return CommandHandlingResult.Fail(TimeSpan.FromMinutes(1));
+                }
+                else
+                {
+                    eventPublisher.PublishEvent(new EthCashinEnrolledToMatchingEngineEvent()
+                    {
+                        CashinOperationId = command.CashinOperationId,
+                        TransactionHash = hash,
+                    });
+                    await _paymentTransactionsRepository.TryCreateAsync(paymentTransaction);
+
+                    return CommandHandlingResult.Ok();
+                }
+            }
+            catch (Exception e)
+            {
+                _log.WriteError(command.TransactionHash ?? "Empty", command, e);
+                throw;
+            }
+        }
+
+        public async Task<CommandHandlingResult> Handle(SaveEthInHistoryCommand command, IEventPublisher eventPublisher)
+        {
+            try
+            {
+                var cashinId = command.CashinOperationId.ToString("N");
+                var clientId = command.ClientId;
+                var hash = command.TransactionHash;
+                var amount = command.Amount;
+                var asset = await _assetsServiceWithCache.TryGetAssetAsync(command.AssetId);
+                var clientAddress = command.ClientAddress;
+
+                var walletCreds = await _walletCredentialsRepository.GetAsync(clientId);
+                await _cashOperationsRepositoryClient.RegisterAsync(new CashInOutOperation
+                {
+                    Id = cashinId,
+                    ClientId = clientId,
+                    Multisig = walletCreds.MultiSig,
+                    AssetId = asset.Id,
+                    Amount = (double)amount,
+                    BlockChainHash = hash,
+                    DateTime = DateTime.UtcNow,
+                    AddressTo = clientAddress,
+                    State = TransactionStates.SettledOnchain
+                });
+
+                var clientAcc = await _clientAccountClient.GetByIdAsync(clientId);
+                await _srvEmailsFacade.SendNoRefundDepositDoneMail(clientAcc.PartnerId, clientAcc.Email, amount, asset.Id);
+                await _paymentTransactionsRepository.SetStatus(hash, PaymentStatus.NotifyProcessed);
```

Comment:
Add ChaosKitty before and after this call

Reply:
fixed